backend/core/utils/tool.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python
+import time
+
+import psycopg2.extras
+
+ISOLEVEL = psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT
+
+LIMIT_RETRIES = 5
+
+
+class DB:
+    def __init__(
+        self,
+        user,
+        password,
+        host,
+        port,
+        database,
+        reconnect=True,
+        autocommit=False,
+        readonly=False,
+    ):
+        self.user = user
+        self.password = password
+        self.host = host
+        self.port = port
+        self.database = database
+        self._connection = None
+        self._cursor = None
+        self.reconnect = reconnect
+        self.autocommit = autocommit
+        self.readonly = readonly
+        self.init()
+
+    def connect(self, retry_counter=0):
+        if not self._connection:
+            try:
+                self._connection = psycopg2.connect(
+                    user=self.user,
+                    password=self.password,
+                    host=self.host,
+                    port=self.port,
+                    database=self.database,
+                    connect_timeout=3,
+                )
+                retry_counter = 0
+                self._connection.set_session(
+                    autocommit=self.autocommit, readonly=self.readonly
+                )
+                return self._connection
+            except psycopg2.OperationalError as error:
+                if not self.reconnect or retry_counter >= LIMIT_RETRIES:
+                    raise error
+                retry_counter += 1
+                print(
+                    "got error {}. reconnecting {}".format(
+                        str(error).strip(), retry_counter
+                    )
+                )
+                time.sleep(5)
+                self.connect(retry_counter)
+            except (Exception, psycopg2.Error) as error:
+                raise error
+
+    def cursor(self):
+        if not self._cursor or self._cursor.closed:
+            if not self._connection:
+                self.connect()
+            self._cursor = self._connection.cursor()
+            return self._cursor
+
+    def execute(self, query, vals=None, retry_counter=0):
```

Comment:
maybe we can have a single function to reuse code? execute, execute_batch and execute_values are more than 90% the same. The difference is in the main call and in the extra page_size param for the values, right?

Reply:
it was more clear for me doing the migration. we can optimize the code afterwards

Reply:
Ok. Resolving.